### ISSUE TITLE
[REFLEXION] Ajout de tests automatiques sur l'API

### ISSUE THREAD

**Issue by jeffd-dev** [jesuisundesdeux/vigilo-backend#265]
Bonjour

Vigilo a été développé pour être rapidement opérationnel du coup il manque de tests automatiques.
Je propose d'en rajouter quelques un au niveau de l'API, afin de s'assurer que le backend est toujours compatible avec les APP (Android...).

Je vois deux avantages à tester au niveau API plutôt que des tests unitaires:
-  on s'assure des réponses de l'API, les tests sont un peu plus lourd mais dépendent pas du code (classes, etc) seront utiles en cas d'évolution et peuvent rester en cas de refonte complète.
- les tests d'API ne dépendent pas d'un language, on peut utiliser un outil (un peu restrictif comme par exemple Postman) mais on peut aussi faire du bash avec curl ou utiliser des scripts classiques PHP ou Python - Pour les possibilité on peut voir [la catégorie Testing de Awesome-Rest-API](https://github.com/Kikobeats/awesome-api#testing) 

Qu'en pensez-vous ?
Avez-vous une préférence sur l'outil ? (J'aurais tendance à expérimenter avec un script python)

**Comment by YvesPasteur** [comment]
Bonjour @jeffd-dev 

C'est une bonne idée.

Pour l'outil, il faut sans doute que ça permette de mocker les dépendances externes pour certains endpoints (par-exemple appels à umap) ?
Dans l'optique de réduire la complexité générale à contribuer, est-ce qu'on ne peut pas plutôt partir sur quelque chose de l'écosystème PHP afin d'éviter de rajouter un langage différent ?

Personnellement pour ce type de tests j'ai déjà eu à utiliser behat (en mode BDD, https://docs.behat.org/en/latest/index.html) ou phpunit (similaire à tous les frameworks de tests XUnit).

**Comment by jeffd-dev** [comment]
Bonjour @YvesPasteur 

- pour umap j'ai pas eu encore besoin de faire de mock, je pense qu'il s'agira d'un besoin pour `getPanel` j'ai aps encore traité ce endpoint
- effectivement je vais rester sur PHP (je pensais a python pour faire un test rapide, mais avec PHP-curl c'est allé presque aussi vite)
- Behat me semble très axé test sur une logique métier importante, là c'est juste de simple appel donc j'ai peur que ça ajoute de la complexité. PhpUnit semble être la bonne référence mais je voulais m'assurer que c'était bon pour tout le monde, parce que ça implique d'utiliser Composer (je sais pas si l'installation du binaire est simple dans un environnement Windows

--  

J'ai commencé à écrire quelques tests en utilisant les assertions/functions de PhpUnit mais en les implémentant moi même pour ne nécessiter aucune dépendance (mais permettre un passage sur PhpUnit ultra simple). 

J'ai crée le OpenApi, c'est une définition normalisée de l'API, en gros la page `doc/REST_API.md` mais en Yaml. Je pensais que l'outil Insomnia allait pouvoir m'aider a créer quelques test rapidement à partir de ce doc mais en fait non. L'avantage d'un document OpenApi c'est qu'il existe des outils (Insomnia, Postman, Swagger ou des extensions Vscode) pour afficher une doc interactive avec la possibilité de faire les appels. C'était pas immédiatement utile ou vraiment nécessaire mais ça m'aide pour les tests.

Il existe toujours un slack ? Est-ce que j'aurais du passer par lui ? J'ai l'impression que le Github est pas très suivi et j'ai plein d'idées à proposer ...

**Comment by YvesPasteur** [comment]
Oui, sur le slack ce sera peut-être plus simple : jesuisundesdeux.slack.com
Il y a un groupe de canaux dédiés à vigilo.
